cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/LineageInfo.java

```diff
@@ -0,0 +1,294 @@
+ * the License.
+ */
+
+package io.cdap.cdap.spi.metadata;
+
+import io.cdap.cdap.api.annotation.Beta;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Lineage information for a program run.
+ */
+@Beta
+public class LineageInfo {
+
+  private final Set<Dataset> sources;
+  private final Set<Dataset> targets;
+  private final Map<Dataset, Set<Dataset>> sourceToTargets;
+  private final Map<Dataset, Set<Dataset>> targetToSources;
+  private final Long startTimeMs;
+  private final Long endTimeMs;
+
+  private LineageInfo(Set<Dataset> sources, Set<Dataset> targets, Map<Dataset, Set<Dataset>> sourceToTargets,
+                      Map<Dataset, Set<Dataset>> targetToSources, @Nullable Long startTimeMs,
+                      @Nullable Long endTimeMs) {
+    this.sources = Collections.unmodifiableSet(new HashSet<>(sources));
+    this.targets = Collections.unmodifiableSet(new HashSet<>(targets));
+    this.sourceToTargets = Collections.unmodifiableMap(new HashMap<>(sourceToTargets));
+    this.targetToSources = Collections.unmodifiableMap(new HashMap<>(targetToSources));
+    this.startTimeMs = startTimeMs;
+    this.endTimeMs = endTimeMs;
+  }
+
+  /**
+   * Returns the set of all source Datasets for a program run.
+   */
+  public Set<Dataset> getSources() {
+    return sources;
+  }
+
+  /**
+   * Returns the set of all target Datasets for a program run.
+   */
+  public Set<Dataset> getTargets() {
+    return targets;
+  }
+
+  /**
+   * Returns the map of all source Datasets to their corresponding set of target Datasets.
+   */
+  public Map<Dataset, Set<Dataset>> getSourceToTargets() {
+    return sourceToTargets;
+  }
+
+  /**
+   * Returns the map of all target Datasets to their corresponding set of source Datasets.
+   */
+  public Map<Dataset, Set<Dataset>> getTargetToSources() {
+    return targetToSources;
+  }
+
+  /**
+   * Returns the time at which lineage was started to get captured.
+   */
+  @Nullable
+  public Long getStartTimeMs() {
+    return startTimeMs;
+  }
+
+  /**
+   * Returns the time at which lineage capture ended.
+   */
+  @Nullable
+  public Long getEndTimeMs() {
+    return endTimeMs;
+  }
+
+  @Override
+  public String toString() {
+    return "LineageInfo{" +
+      "sources=" + sources +
+      ", targets=" + targets +
+      ", sourceToTargets=" + sourceToTargets +
+      ", targetToSources=" + targetToSources +
+      ", startTimeMs=" + startTimeMs +
+      ", endTimeMs=" + endTimeMs +
+      '}';
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    LineageInfo that = (LineageInfo) o;
+    return getSources().equals(that.getSources()) && getTargets().equals(that.getTargets()) &&
```

Comment:
It is always better to use `Objects.equals` so you don't need to worry about null.  Also no need to use `get` for the variable.